WindowsPipeReceiver/Sender inherit a pure virtual function but do not override it

https://github.com/weidai11/cryptopp/blob/48809d4e85c125814425c621d8d0d89f95405924/winpipes.h#L60

They inherit `virtual HANDLE GetHandle() const =0;` from `WindowsPipe`.

It is then called in `WindowsPipeReceiver::Receive()`, `WindowsPipeReceiver::GetReceiveResult()`, `WindowsPipeSender::Send` and `WindowsPipeSender::GetSendResult()`:
https://github.com/weidai11/cryptopp/blob/48809d4e85c125814425c621d8d0d89f95405924/winpipes.cpp#L94

At this state: https://github.com/weidai11/cryptopp/tree/8a02fe1787e62fb14f90650dee57024920c83adb
Compilation of `cryptest.exe` failed because of this:
```
g++ -o cryptest.exe -DNDEBUG -g2 -O3 -fPIC -march=native -Wall -Wextra -Wno-type-limits -Wno-unknown-pragmas -pipe bench.o bench2.o test.o validat0.o validat1.o validat2.o validat3.o adhoc.o datatest.o regtest.o fipsalgt.o dlltest.o ./libcryptopp.a  -lws2_32
./libcryptopp.a(winpipes.o): In function `ZN8CryptoPP19WindowsPipeReceiver16GetReceiveResultEv':
C:\GitHub\cryptopp/winpipes.cpp:131: undefined reference to `non-virtual thunk to CryptoPP::WindowsPipeSource::GetHandle() const'
C:\GitHub\cryptopp/winpipes.cpp:131: undefined reference to `non-virtual thunk to CryptoPP::WindowsPipeSource::GetHandle() const'
./libcryptopp.a(winpipes.o): In function `ZN8CryptoPP19WindowsPipeReceiver7ReceiveEPhj':
C:\GitHub\cryptopp/winpipes.cpp:94: undefined reference to `non-virtual thunk to CryptoPP::WindowsPipeSource::GetHandle() const'
C:\GitHub\cryptopp/winpipes.cpp:94: undefined reference to `non-virtual thunk to CryptoPP::WindowsPipeSource::GetHandle() const'
collect2.exe: error: ld returned 1 exit status
```

After you pushed 5.6.3 RC6, these errors do not appear anymore. But I guess that's because these classes are not tested? Because you still have a pure virtual function that isn't overridden.

--- a/winpipes.cpp
+++ b/winpipes.cpp
@@ -91,7 +91,7 @@ bool WindowsPipeReceiver::Receive(byte* buf, size_t bufLen)
 {
 	assert(!m_resultPending && !m_eofReceived);
 
-	HANDLE h = GetHandle();
+	const HANDLE h = GetHandle();
 	// don't queue too much at once, or we might use up non-paged memory
 	if (ReadFile(h, buf, UnsignedMin((DWORD)128*1024, bufLen), &m_lastResult, &m_overlapped))
 	{
@@ -128,7 +128,7 @@ unsigned int WindowsPipeReceiver::GetReceiveResult()
 {
 	if (m_resultPending)
 	{
-		HANDLE h = GetHandle();
+		const HANDLE h = GetHandle();
 		if (GetOverlappedResult(h, &m_overlapped, &m_lastResult, false))
 		{
 			if (m_lastResult == 0)
@@ -165,7 +165,7 @@ WindowsPipeSender::WindowsPipeSender()
 void WindowsPipeSender::Send(const byte* buf, size_t bufLen)
 {
 	DWORD written = 0;
-	HANDLE h = GetHandle();
+	const HANDLE h = GetHandle();
 	// don't queue too much at once, or we might use up non-paged memory
 	if (WriteFile(h, buf, UnsignedMin((DWORD)128*1024, bufLen), &written, &m_overlapped))
 	{
@@ -193,7 +193,7 @@ unsigned int WindowsPipeSender::GetSendResult()
 {
 	if (m_resultPending)
 	{
-		HANDLE h = GetHandle();
+		const HANDLE h = GetHandle();
 		BOOL result = GetOverlappedResult(h, &m_overlapped, &m_lastResult, false);
 		CheckAndHandleError("GetOverlappedResult", result);
 		m_resultPending = false;

--- a/winpipes.h
+++ b/winpipes.h
@@ -21,7 +21,7 @@ public:
 	bool GetOwnership() const {return m_own;}
 	void SetOwnership(bool own) {m_own = own;}
 
-	operator HANDLE() {return m_h;}
+	operator HANDLE() const {return m_h;}
 	HANDLE GetHandle() const {return m_h;}
 	bool HandleValid() const;
 	void AttachHandle(HANDLE h, bool own=false);
@@ -67,6 +67,7 @@ public:
 	unsigned int GetReceiveResult();
 	bool EofReceived() const {return m_eofReceived;}
 
+	HANDLE GetHandle() const {return m_event;}
 	unsigned int GetMaxWaitObjectCount() const {return 1;}
 	void GetWaitObjects(WaitObjectContainer &container, CallStack const& callStack);
 
@@ -90,6 +91,7 @@ public:
 	bool MustWaitForEof() { return false; }
 	void SendEof() {}
 
+	HANDLE GetHandle() const {return m_event;}
 	unsigned int GetMaxWaitObjectCount() const {return 1;}
 	void GetWaitObjects(WaitObjectContainer &container, CallStack const& callStack);
 

